cpp/src/arrow/types/struct.cc

```diff
@@ -73,10 +76,21 @@ Status StructArray::Validate() const {
     }
 
     // Validate null bitmap
-    // TODO: No checks for the consistency of bitmaps. Its cost maybe expensive.
     if (array_length > 0 && array_length != length_) {
       return Status::Invalid("Struct's length is not equal to its child arrays");
     }
+    bool is_null = true;
+    for (int32_t i = 0; i < length_; ++i) {
+      if (IsNull(i)) continue;
+      for (auto arr : field_arrays_) {
+        is_null &= arr->IsNull(i);
+      }
+      if (is_null) {
+        std::stringstream ss;
+        ss << "Slot is invalid at pos " << i << ", but indicated as valid by bitmap";
+        return Status::Invalid(ss.str());
+      }
+    }
```

Comment:
@wesm Done. I think I've misunderstood some comments.
Then other improvements are under cover? I will refine it instantly.
Thanks.